Source/JavaScriptCore/heap/FreeList.cpp

```diff
@@ -39,53 +39,50 @@ FreeList::~FreeList()
 
 void FreeList::clear()
 {
-    m_scrambledHead = 0;
+    m_head = reinterpret_cast_ptr<FreeCell*>(1);
     m_secret = 0;
+    m_payloadStart = nullptr;
     m_payloadEnd = nullptr;
-    m_remaining = 0;
     m_originalSize = 0;
 }
 
-void FreeList::initializeList(FreeCell* head, uintptr_t secret, unsigned bytes)
+void FreeList::initialize(FreeCell* start, uint64_t secret, unsigned bytes)
 {
-    // It's *slightly* more optimal to use a scrambled head. It saves a register on the fast path.
-    m_scrambledHead = FreeCell::scramble(head, secret);
+    if (UNLIKELY(!start)) {
+        clear();
+        return;
+    }
+    auto [offsetToNext, lengthInBytes] = start->decode(secret);
+    m_head = bitwise_cast<FreeCell*>(bitwise_cast<char*>(start) + offsetToNext);
     m_secret = secret;
-    m_payloadEnd = nullptr;
-    m_remaining = 0;
+    m_payloadStart = reinterpret_cast<char*>(start);
```

Comment:
Use `bitwise_cast` here instead of `reinterpret_cast`.  You can also initialize `m_head` in terms of `m_payloadStart` instead of having to do `bitwise_cast<char*>(start)` 2 times (not that it really matters for perf, but I think it's good to clearly point out the the head is an offset from the start of the current payload / interval.

Reply:
Resolved in latest push.